src/util/bson/bson.cpp

```diff
@@ -0,0 +1,1164 @@
+
+    bool parse_error(std::size_t,
+                     const std::string&,
+                     const nlohmann::detail::exception& ex) override;
+
+    Bson parse(const std::string& json);
+
+protected:
+    struct Instruction {
+        State type;
+        std::string key;
+    };
+
+    /// Convenience class that overloads similar collection functionality
+    class BsonContainer {
+        typedef enum Type {
+            DOCUMENT, ARRAY
+        } Type;
+        Type m_type;
+
+        union {
+            BsonDocument* document;
+            BsonArray* array;
+        };
+    public:
+        ~BsonContainer() noexcept
+        {
+            if (m_type == DOCUMENT) {
+                delete document;
+                document = NULL;
+            } else {
+                delete array;
+                array = NULL;
+            }
+        }
+
+        BsonContainer(const BsonDocument& v) noexcept : document(new BsonDocument(v)) {
+            m_type = DOCUMENT;
+        }
+        BsonContainer(const BsonArray& v) noexcept : array(new BsonArray(v)) {
+            m_type = ARRAY;
+        }
+        BsonContainer(BsonDocument&& v) noexcept :
+            m_type(DOCUMENT), document(new BsonDocument(std::move(v))) {}
+
+        BsonContainer(BsonArray&& v) noexcept : m_type(ARRAY), array(new BsonArray(std::move(v))) {}
+
+        explicit operator const BsonDocument&&() const {
+            REALM_ASSERT(is_document());
+            return std::move(*document);
+        }
+
+        explicit operator const BsonArray&&() const {
+            REALM_ASSERT(is_array());
+            return std::move(*array);
+        }
+
+        BsonContainer(const BsonContainer& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+        }
+
+        BsonContainer& operator=(const BsonContainer& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+            return *this;
+        }
+
+        BsonContainer& operator=(BsonContainer&& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                if (document) delete document;
+                document = v.document;
+                v.document = NULL;
+            } else {
+                if (array) delete array;
+                array = v.array;
+                v.array = NULL;
+            }
+            return *this;
+        }
+
+        BsonContainer(BsonContainer&& v) {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                if (document) delete document;
```

Comment:
No need to clean up memory in the constructor because we haven't allocated anything yet. Also don't forget that on linux, pointers aren't zero initialized so this will go and delete random memory somewhere and usually crash.